crates/proof-of-sql/src/sql/proof_exprs/range_check.rs

```diff
@@ -0,0 +1,412 @@
+            .collect();
+
+        // Invert all the terms in a row at once
+        // inverted_word_columns:
+        //
+        // | Column i            | Column i+1            | Column i+2            | ... | Column_||word||     |
+        // |---------------------|-----------------------|-----------------------|-----|---------------------|
+        // | 1/(word[i] + alpha) | 1/(word[i+1] + alpha) | 1/(word[i+2] + alpha) | ... | 1/(word[n] + alpha) |
+        slice_ops::batch_inversion(&mut terms_to_invert);
+
+        // Assign the inverted values back to the inverted_word_columns
+        for ((j, &inverted_value), column) in terms_to_invert
+            .iter()
+            .enumerate()
+            .zip(inverted_word_columns.iter_mut())
+        {
+            column[i] = inverted_value; // j is the column index, i is the row index
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        base::scalar::{Curve25519Scalar as S, Scalar},
+        sql::proof_exprs::range_check::{decompose_scalar_to_words, get_logarithmic_derivative},
+    };
+    use bumpalo::Bump;
+    use bytemuck;
+    use num_traits::Inv;
+    use rand::Rng;
+
+    #[test]
+    fn we_can_decompose_large_scalars_to_words() {
+        let mut scalars: Vec<S> = [S::MAX_SIGNED, S::from(u64::MAX), S::from(-1)]
+            .iter()
+            .map(S::from)
+            .collect();
+
+        let alloc = Bump::new();
+        let mut word_columns: Vec<&mut [u8]> = (0..31)
+            .map(|_| alloc.alloc_slice_fill_with(scalars.len(), |_| 0u8))
+            .collect();
+
+        let byte_counts = alloc.alloc_slice_fill_with(256, |_| 0u64);
+        let mut all_scalar_bytes: Vec<&[u8]> = Vec::with_capacity(scalars.len());
+
+        decompose_scalar_to_words(
+            &mut scalars,
+            &alloc,
+            &mut word_columns,
+            byte_counts,
+            &mut all_scalar_bytes,
+        );
+
+        // Word representation of S::MAX_SIGNED
+        let res_1: Vec<u8> = vec![
+            246, 233, 122, 46, 141, 49, 9, 44, 107, 206, 123, 81, 239, 124, 111, 10, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 8,
+        ];
+        assert_eq!(all_scalar_bytes[0], &res_1);
+
+        // Word representation of u64::MAX
+        let res_2: Vec<u8> = vec![
+            255, 255, 255, 255, 255, 255, 255, 255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0,
+        ];
+        assert_eq!(all_scalar_bytes[1], &res_2);
+
+        // Word representation of S::from(-1) aka P -1
+        let res_3: Vec<u8> = vec![
+            236, 211, 245, 92, 26, 99, 18, 88, 214, 156, 247, 162, 222, 249, 222, 20, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 16,
+        ];
+
+        assert_eq!(all_scalar_bytes[2], &res_3);
+    }
+
+    #[test]
+    fn we_can_decompose_small_scalars_to_words() {
+        let mut scalars: Vec<S> = [1, 2, 3].iter().map(S::from).collect();
```

Comment:
The decomposition of this column into columns of words should be
`vec![1,2,3], vec![0,0,0], vec![0,0,0], ..., vec![0,0,0]`.